src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -32,6 +32,9 @@ end
 #####
 
 function rrule(::Type{<:Diagonal}, d::AbstractVector)
+    function Diagonal_pullback(x::AbstractThunk)
+        return Diagonal_pullback(unthunk(x))
+    end
     function Diagonal_pullback(ȳ::AbstractMatrix)
```

Comment:
I'm just wondering how did that work before? I.e. would there not be a `MethodError` if an `AbstractThunk` is passed to this? Or is there some black magic in how thunks work?